src/analysis/convex/extreme.lean

```diff
@@ -175,13 +168,61 @@ subset.antisymm (λ x hx, ⟨hx.1, hAB.extreme_points_subset_extreme_points hx
 end has_smul
 
 section ordered_semiring
-variables {𝕜} [ordered_semiring 𝕜] [add_comm_group E] [module 𝕜 E] {A B : set E} {x : E}
+variables [ordered_semiring 𝕜] [add_comm_group E] [add_comm_group F] [Π i, add_comm_group (π i)]
+  [module 𝕜 E] [module 𝕜 F] [Π i, module 𝕜 (π i)] {A B : set E} {x : E}
 
 lemma is_extreme.convex_diff (hA : convex 𝕜 A) (hAB : is_extreme 𝕜 A B) :
   convex 𝕜 (A \ B) :=
 convex_iff_open_segment_subset.2 (λ x₁ ⟨hx₁A, hx₁B⟩ x₂ ⟨hx₂A, hx₂B⟩ x hx,
     ⟨hA.open_segment_subset hx₁A hx₂A hx, λ hxB, hx₁B (hAB.2 hx₁A hx₂A hxB hx).1⟩)
 
+@[simp] lemma extreme_points_prod (s : set E) (t : set F) :
+  (s ×ˢ t).extreme_points 𝕜 = s.extreme_points 𝕜 ×ˢ t.extreme_points 𝕜 :=
+begin
+  ext,
+  refine (and_congr_right $ λ hx, ⟨λ h, _, λ h, _⟩).trans (and_and_and_comm _ _ _ _),
+  split,
```

Comment:
I think this lemma would be helpful here:
```lean
lemma open_segment_mk_left (x₁ x₂ : E) (y : F) :
  open_segment 𝕜 (x₁, y) (x₂, y) = (λ x, (x, y)) '' open_segment 𝕜 x₁ x₂ :=
begin
  ext ⟨x', y'⟩,
  simp_rw [set.mem_image, open_segment, set.mem_set_of, prod.smul_mk, prod.mk_add_mk,
    prod.mk.inj_iff],
  split,
  { rintros ⟨a, b, ha, hb, hab, hx⟩,
    simp_rw [convex.combo_self hab] at hx,
    refine ⟨x', ⟨a, b, ha, hb, hab, hx.1⟩, rfl, hx.2⟩, },
  { rintros ⟨x, ⟨a, b, ha, hb, hab, hx⟩, rfl, rfl⟩,
    refine ⟨a, b, ha, hb, hab, hx, convex.combo_self hab _⟩ }
end
```
There would be a much shorter proof if `open_segment_image` were generalized to affine maps.

Reply:
Unfortunately we can't use those generalised `segment_image`/`open_segment_image` because affinity is not a generalisation of convexity (and in particular an `ordered_semiring` is not a `ring`).

Reply:
I've added your lemmas, but I can't see how they shorten the proof. Everything is under a lot of binders and manipulation is hard.

Reply:
I'll have a go now